varcode/variant.py

```diff
@@ -46,3 +114,240 @@ def short_description(self):
+        if not is_coding_biotype(transcript.biotype):
+            return NoncodingTranscript(self, transcript)
+
+        if not transcript.complete:
+            return IncompleteTranscript(self, transcript)
+
+        distance_to_exon, nearest_exon = find_nearest_locus(
+            start=self.pos,
+            end=self.end,
+            loci=transcript.exons)
+
+        if distance_to_exon > 0:
+            return self.intronic_transcript_effect(
+                transcript=transcript,
+                nearest_exon=nearest_exon)
+
+        # TODO: exonic splice site mutations
+        return self.exonic_transcript_effect(transcript)
+
+
+    def intronic_transcript_effect(self, transcript, nearest_exon):
+        """
+        Infer effect of variant which does not overlap any exon of
+        the given transcript.
+        """
+        distance_to_exon = nearest_exon.distance_to_interval(
+            self.pos, self.end)
+
+        assert distance_to_exon > 0, \
+            "Expected intronic effect to have distance_to_exon > 0, got %d" % (
+                distance_to_exon,)
+
+        before_forward_exon = (
+            transcript.strand == "+" and
+            self.pos < nearest_exon.start)
+
+        before_backward_exon = (
+            transcript.strand == "-" and
+            self.end > nearest_exon.end)
+
+        before_exon = before_forward_exon or before_backward_exon
+
+        if distance_to_exon <= 2:
+            if before_exon:
+                # 2 last nucleotides of intron before exon are the splice acceptor
+                # site, typically "AG"
+                effect_class = SpliceAcceptor
+            else:
+                # 2 first nucleotides of intron after exon are the splice donor
+                # site, typically "GT"
+                effect_class = SpliceDonor
+        elif not before_exon and distance_to_exon <= 6:
+            # variants in nucleotides 3-6 at start of intron aren't as certain
+            # to cause problems as nucleotides 1-2 but still implicated in
+            # alternative splicing
+            effect_class = IntronicSpliceSite
+        elif before_exon and distance_to_exon <= 4:
+            # nucleotides -4 and -3 before exon are part of the 3' splicing motif
+            # but allow for more degeneracy than the -2, -1 nucleotides
+            effect_class = IntronicSpliceSite
+        else:
+            assert distance_to_exon > 6, \
+                "Looks like we didn't cover all possible splice site mutations"
+            # intronic mutation unrelated to splicing
+            effect_class = Intronic
+
+        return effect_class(
+                variant=self,
+                transcript=transcript,
+                nearest_exon=nearest_exon,
+                distance_to_exon=distance_to_exon)
+
+
+    def offset_on_transcript(self, transcript):
+        """
+        Given a Variant (with fields `ref`, `alt`, `pos`, and `end`),
```

Comment:
suggested minor cleanup:

Given a Variant (with fields `ref`, `alt`, `pos`, and `end`), compute the offset of the variant position relative to the position and direction of the transcript. Returns a tuple (offset, ref, alt), where `ref` and `alt` are this variant's `ref` and `alt` nucleotide strings with shared prefixes and suffixes removed and the bases reversed if the transcript is on the backward strand.

For example,

```
>> variant = Variant(ref="ATGGC", alt="TCGC", pos=3000, end=3005)
>> variant.offset_on_transcript(Transcript(start=2000, end=4000))
(1002, "G", "C")
```